Sources/_MatchingEngine/Engine/Engine.swift

```diff
@@ -1,6 +1,6 @@
 // Currently, engine binds the type and consume binds an instance.
 // But, we can play around with this.
-public struct Engine<Input: Collection> where Input.Element: Hashable {
+public struct Engine<Input: BidirectionalCollection> where Input.Element: Hashable {
```

Comment:
(please ignore until you're back from break) 

@rxwei, I had to make this Bidi for easier anchor or assertions like `\b`. I still want to make some kind of capabilities system where you can use ME on non-bidi or even non-collections provided you don't use those instructions, but that will probably have to be added later.